### PR TITLE
Supporting Date,timestamp ,Decimal,Struct, Array and Map data types.

### PR DESCRIPTION
**Pull request by ravipesala**
Following data types are supported in Astro with this PR. Now Astro can support all data types which are supported Astro.
1. Date
2.Timestamp
3.Decimal
4.Struct
5.Array
6.Map

And refactored code to get good maintainability and support complex datatypes.

### CODE REVIEW

**Comment by ravipesala** [comment]
ok to test

**Comment by ravipesala** [comment]
retest this please

**Comment by ravipesala** [comment]
@jackylk @WangTaoTheTonic @scwf Please review this PR.

**Comment by ravipesala** [comment]
retest this please

**Comment by sisihj** [comment]
@Sephiroth-Lin, pls review this PR.

File: src/main/scala/org/apache/spark/sql/hbase/HBaseRelation.scala [hunk 1]
```diff
@@ -50,7 +51,7 @@ class HBaseSource extends SchemaRelationProvider {
     val optSet = Set("hbasetablename", "keycols", "colsmapping")
     val keySet = parameters.keySet.map(_.toLowerCase)
     if (!(optSet.intersect(keySet).size == 3)) {
-      throw new Exception("Options keys(hbaseTableName, keyCols, colsMapping) are obligatory")
+      throw new scala.Exception("Options keys(hbaseTableName, keyCols, colsMapping) are obligatory")
```
Comment (Sephiroth-Lin) [inline src/main/scala/org/apache/spark/sql/hbase/HBaseRelation.scala]: May be we can use SparkException or AstroException, scala.Exception is same as Exception, right?

File: src/main/scala/org/apache/spark/sql/hbase/dataTypeReadWriteObjects.scala [hunk 2]
```diff
@@ -0,0 +1,1092 @@
...
+  }
+
+  def getDataStorageFormat: String = FieldFactory.BINARY_FORMAT
+}
+
+case class ArrayDataField(dataType: ArrayType,
+                          encodingFormat: String,
+                          separators: Array[Byte],
+                          level: Int,
+                           hiveStorage: Boolean) extends FieldData {
```
Comment (Sephiroth-Lin) [inline src/main/scala/org/apache/spark/sql/hbase/dataTypeReadWriteObjects.scala]: code style

File: src/main/scala/org/apache/spark/sql/hbase/util/DataTypeUtils.scala [hunk 3]
```diff
@@ -34,188 +34,35 @@ import org.apache.spark.unsafe.types.UTF8String
...
-      case IntegerType => bu.toBytes(row.getInt(index))
-      case LongType => bu.toBytes(row.getLong(index))
-      case ShortType => bu.toBytes(row.getShort(index))
-      case StringType => bu.toBytes(row.getString(index))
-      case DateType => bu.toBytes(row.getInt(index))
-      case TimestampType => bu.toBytes(row.getLong(index))
-      case _ => throw new Exception("Unsupported HBase SQL Data Type")
-    }
-  }
+    :: DoubleType :: FloatType :: DateType :: TimestampType :: DecimalType :: Nil).toSet
```
Comment (Sephiroth-Lin) [inline src/main/scala/org/apache/spark/sql/hbase/util/DataTypeUtils.scala]: Now we can support array/map/struct right?

**Comment by ravipesala** [comment]
Updated as per comments. Please review

**Comment by ravipesala** [comment]
Pls review, if everything is fine I will merge this PR

**Comment by Sephiroth-Lin** [comment]
I think is fine. @jackylk @scwf @WangTaoTheTonic What's your opinion?

**Comment by WangTaoTheTonic** [comment]
LGTM +1

**Comment by ravipesala** [comment]
Thank you, Merged to spark1.5-alpha-ravi

**Pull request merged**